File: examples/gradle-ktor/src/main/kotlin/community/flock/wirespec/example/maven/custom/app/todo/TodoService.kt
package community.flock.wirespec.example.maven.custom.app.todo

interface TodoService : HasTodoRepository

fun TodoService.getAllTodos() = todoRepository.getAllTodos()

fun TodoService.getTodoById(id: Todo.Id) = todoRepository.getTodoById(id)

fun TodoService.saveTodo(todo: Todo) = todoRepository.saveTodo(todo)

fun TodoService.deleteTodoById(id: Todo.Id) = todoRepository.deleteTodoById(id)

interface HasTodoService {
    val todoService: TodoService
}

File: examples/scala-zio-2/src/main/scala/example/service/TodoService.scala
package example.service

import community.flock.wirespec.generated.model.{PotentialTodoDto, TodoDto, TodoId}
import example.port.{TodoRepository, TodoService}
import zio.*

import java.util.UUID

class TodoServiceLive(repo: TodoRepository) extends TodoService:

  override def getTodos(done: Option[Boolean]): Task[List[TodoDto]] =
    repo.findAll().map(_.filter(t => done.forall(_ == t.done)))

  override def createTodo(potential: PotentialTodoDto): Task[TodoDto] =
    for
      id  <- ZIO.succeed(TodoId(UUID.randomUUID().toString))
      dto  = TodoDto(
               id       = id,
               name     = potential.name,
               done     = potential.done,
               testInt0 = potential.testInt0,
               testInt1 = potential.testInt1,
               testInt2 = potential.testInt2,
               testNum0 = potential.testNum0,
               testNum1 = potential.testNum1,
               testNum2 = potential.testNum2,
             )
      saved <- repo.save(dto)
    yield saved

  override def getTodoById(id: TodoId): Task[Option[TodoDto]] =
    repo.findById(id)

  override def deleteTodoById(id: TodoId): Task[Option[TodoDto]] =
    repo.delete(id)

object TodoServiceLive:
  val layer: URLayer[TodoRepository, TodoService] =
    ZLayer.fromZIO(ZIO.service[TodoRepository].map(new TodoServiceLive(_)))
